File: src/mbs/application/article/create.py
from datetime import datetime
from typing import Protocol

from pydantic import BaseModel

import mbs.domain.exceptions as domain_exceptions
from mbs.application.common.article_gateway import ArticleWriter
from mbs.application.common.exceptions import Unauthorized, InvalidData, Forbidden
from mbs.application.common.id_provider import IdProvider
from mbs.application.common.interactor import Interactor
from mbs.application.common.tag_gateway import TagReader, TagLinker, TagWriter
from mbs.domain.models import ArticleState, ArticleId, FileId, UserId
from mbs.domain.services.access import AccessService
from mbs.domain.services.article import ArticleService
from mbs.domain.services.tag import TagService
from mbs.domain.services.validator import ValidatorService


class CreateArticleDTO(BaseModel):
    title: str
    content: str
    state: ArticleState
    tags: list[str]


class CreateArticleResult(BaseModel):
    id: ArticleId
    title: str
    content: str
    poster: FileId | None
    views: int
    likes: int
    tags: list[str]
    state: ArticleState
    author_id: UserId

    created_at: datetime
    updated_at: datetime | None


class TagGateway(TagReader, TagWriter, TagLinker, Protocol):
    pass


class CreateArticle(Interactor[CreateArticleDTO, CreateArticleResult]):

    def __init__(
            self,
            article_writer: ArticleWriter,
            access_service: AccessService,
            article_service: ArticleService,
            tag_gateway: TagGateway,
            tag_service: TagService,
            validator: ValidatorService,
            id_provider: IdProvider,
    ):
        self._article_writer = article_writer
        self._access_service = access_service
        self._article_service = article_service
        self._tag_gateway = tag_gateway
        self._tag_service = tag_service
        self._validator = validator
        self._id_provider = id_provider

    async def __call__(self, data: CreateArticleDTO) -> CreateArticleResult:

        try:
            self._access_service.ensure_can_create_article(
                is_auth=self._id_provider.is_auth(),
                permissions=self._id_provider.permissions(),
                user_state=self._id_provider.user_state(),
            )
        except domain_exceptions.AuthenticationError as error:
            raise Unauthorized(str(error))
        except domain_exceptions.AccessDenied as error:
            raise Forbidden(str(error))

        validator_err_map: dict[str, str] = dict()
        try:
            self._validator.validate_article_title(data.title)
        except domain_exceptions.ValidationError as error:
            validator_err_map["title"] = str(error)

        try:
            self._validator.validate_article_content(data.content)
        except domain_exceptions.ValidationError as error:
            validator_err_map["content"] = str(error)

        try:
            for tag in data.tags:
                self._validator.validate_tag_title(tag)
        except domain_exceptions.ValidationError as error:
            validator_err_map["tags"] = str(error)

        if validator_err_map:
            raise InvalidData(validator_err_map)

        article = self._article_service.create_article(
            title=data.title,
            content=data.content,
            state=data.state,
            author_id=self._id_provider.user_id(),
            tags=data.tags
        )
        await self._article_writer.save_article(article)

        return CreateArticleResult(
            id=article.id,
            title=article.title,
            content=article.content,
            poster=article.poster,
            views=article.views,
            likes=article.likes,
            tags=article.tags,
            state=article.state,
            author_id=article.author_id,
            created_at=article.created_at,
            updated_at=article.updated_at
        )

File: src/application/article/create.rs
use crate::application::common::{
    article_gateway::ArticleWriter,
    error::AppError,
    id_provider::IdProvider,
    interactor::Interactor
};
use crate::domain::error::ValidationError;
use crate::domain::models::tag::Tag;
use crate::domain::{
    error::DomainError,
    models::{
        article::{Article, ArticleId},
        article_state::ArticleState
    },
    services::{
        access::ensure_can_create_article,
        validator::{
            validate_article_content,
            validate_article_tags,
            validate_article_title
        }
    }
};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use utoipa::ToSchema;

#[derive(Deserialize, ToSchema)]
pub struct CreateArticleInput {
    #[schema(example = "Super rust tips")]
    pub title: String,
    #[schema(example = "In this article, we will explore some of the best practices in Rust programming...")]
    pub content: String,
    /// When creating, it is better to use Draft
    #[schema(example = "Draft")]
    pub state: ArticleState,
    #[schema(example = json!(vec!["rust", "programming", "tips"]), value_type = Vec<String>)]
    pub tags: Vec<String>
}

#[derive(Serialize, ToSchema)]
pub struct CreateArticleOutput {
    #[schema(example = uuid::Uuid::new_v4, value_type=uuid::Uuid)]
    pub id: ArticleId
}

pub struct CreateArticle<'interactor> {
    pub id_provider: Box<dyn IdProvider>,
    pub article_writer: &'interactor dyn ArticleWriter
}

impl Interactor<CreateArticleInput, CreateArticleOutput> for CreateArticle<'_> {
    async fn execute(&self, input: CreateArticleInput) -> Result<CreateArticleOutput, AppError> {
        ensure_can_create_article(
            self.id_provider.permissions(),
            self.id_provider.user_state(),
        )?;

        // validate
        let mut validator_err_map = HashMap::<String, ValidationError>::new();
        if let Err(DomainError::Validation((key, val))) = validate_article_title(&input.title) {
            validator_err_map.insert(key, val);
        }
        if let Err(DomainError::Validation((key, val))) = validate_article_content(&input.content) {
            validator_err_map.insert(key, val);
        }
        if let Err(DomainError::Validation((key, val))) = validate_article_tags(&input.tags) {
            validator_err_map.insert(key, val);
        }
        if !validator_err_map.is_empty() {
            return Err(AppError::Validation(validator_err_map));
        }
        
        // save
        let article = Article::new(
            input.title,
            None,
            input.content,
            input.tags.into_iter().map(|title| Tag::new(title)).collect(),
            input.state,
            *self.id_provider.user_id(),
        );

        self.article_writer.save(&article).await?;
        
        Ok(CreateArticleOutput { id: article.id })
    }
}
